In [8]:
import numpy as np

In [9]:
#1. Generate statistics
np.random.seed(42) #reproducibility
consumption = np.random.randint(200, 600, size=120)
print("Consumption:", consumption)

Consumption: [302 548 470 306 271 388 220 302 321 414 530 287 572 299 559 351 330 349
 508 457 543 493 585 391 476 360 513 221 452 435 544 248 258 369 387 470
 389 374 250 563 254 443 519 330 506 334 220 528 366 473 587 288 515 213
 441 464 545 252 585 539 291 566 463 234 405 280 249 559 587 201 589 253
 305 459 509 390 417 243 361 401 469 550 503 470 414 451 389 495 412 407
 436 537 566 252 479 416 451 387 579 240 356 214 500 264 544 526 208 543
 328 335 262 338 280 591 362 488 578 460 430 240]


In [17]:
#2. Basic statistics
total_consumption = np.sum(consumption)
mean_consumption = np.mean(consumption)
median_consumption = np.median(consumption)
min_consumption = np.min(consumption)
max_consumption = np.max(consumption)
std_consumption = np.std(consumption) 

print("Total consumption:", total_consumption)
print("Average consumption:", (round(mean_consumption, 2)))
print("Median consumption:", median_consumption)
print("Minimun consumption:", min_consumption)
print("Maximum consumption:", max_consumption)
print("Standard deviation:", (round(std_consumption, 2)))

Total consumption: 49019
Average consumption: 408.49
Median consumption: 414.0
Minimun consumption: 201
Maximum consumption: 591
Standard deviation: 115.2


In [11]:
#3. Identify extremes
day_highest_consumption = np.where(consumption == np.max(consumption))[0] + 1
day_lowest_consumption = np.where(consumption == np.min(consumption))[0] + 1

print("Day of highest consumtion:", day_highest_consumption)
print("Day of lowest consumtion:", day_lowest_consumption)

Day of highest consumtion: [114]
Day of lowest consumtion: [70]


In [12]:
#4. Monthly spliting
months = np.array_split(consumption, 4)
each_month = [print(month) for month in months]

[302 548 470 306 271 388 220 302 321 414 530 287 572 299 559 351 330 349
 508 457 543 493 585 391 476 360 513 221 452 435]
[544 248 258 369 387 470 389 374 250 563 254 443 519 330 506 334 220 528
 366 473 587 288 515 213 441 464 545 252 585 539]
[291 566 463 234 405 280 249 559 587 201 589 253 305 459 509 390 417 243
 361 401 469 550 503 470 414 451 389 495 412 407]
[436 537 566 252 479 416 451 387 579 240 356 214 500 264 544 526 208 543
 328 335 262 338 280 591 362 488 578 460 430 240]


In [13]:
#5. Monthly Analysis
print("Monthly Total")
monthly_total = [print(month.sum()) for month in months]

print("\nMonthly Average")
monthly_avg = [print(round(month.mean(), 2)) for month in months]

Monthly Total
12253
12254
12322
12190

Monthly Average
408.43
408.47
410.73
406.33


In [19]:
#6. Weekly Analysis
each_week = [month[:28].reshape(4, 7) for month in months]

print("Weekly Total")
weekly_total = [print(week.sum(axis=1)) for week in each_week]

print("\nWeekly Average")
weekly_avg = [print(week.mean(axis=1)) for week in each_week]

Weekly Total
[2505 2725 3097 3039]
[2665 2733 3014 2718]
[2488 2953 2790 3272]
[3137 2540 2746 3097]

Weekly Average
[357.85714286 389.28571429 442.42857143 434.14285714]
[380.71428571 390.42857143 430.57142857 388.28571429]
[355.42857143 421.85714286 398.57142857 467.42857143]
[448.14285714 362.85714286 392.28571429 442.42857143]


In [15]:
#7. Overall Consumption
cumulative_consumption = np.cumsum(consumption)

print("First 10 cumulative consumption values:")
print(cumulative_consumption[:10])

First 10 cumulative consumption values:
[ 302  850 1320 1626 1897 2285 2505 2807 3128 3542]


In [16]:
#8. Anomaly Detection
lower_bound = mean_consumption - 2 * std_consumption
upper_bound = mean_consumption + 2 * std_consumption

print(round(lower_bound, 2))
print(round(upper_bound, 2))

anomaly = np.where((consumption < lower_bound) | (consumption > upper_bound)) [0] + 1
print("Anomaly detection (low/high):", anomaly)

178.09
638.9
Anomaly detection (low/high): []


*Summary of Electricity Consumption Analysis* ---
The analysis showed that the total consumption over the four-month period is "49019kwh." The Average consumption of the entire period is 414.0Kwh, whereby the maximum and minimum consumption are 591Kwh and 201Kwh respescctively.The days with the higest and lowest consumption is **Day 70** and **Day 114**, respectively. The monthly total of the consumption are 12253Kwh, 12254Kwh, 12322Kwh, 12190Kwh, with the third month having the highest consumption of 12322Kwh. The weekly analysis shows that in the 1st and 2nd month, week 3 has the highest average consumption of 442.43Kwh and 430.57Kwh, respectively. In 3rd and 4th month, week 4 and week 1 had the highest average producution of 467.43Kwh and 448.14Kwh, respectively. The analysis shows that there are no anomalies in the data as the highest and lowest consumptions are with the range of the upper and lower bond respectively. 